File: C_plus_plus/Find_Minimum_in_Rotated_Sorted_Array_II.cpp
class Solution {
public:
    int findMin(vector<int> &num) {
        int i = 0, j = num.size()-1,k;
        while(i<=j){
            if(i==j||num[i]<num[j]) return num[i];
            k = (i+j)/2;
            if(k+1<=j&&num[k+1]<num[k]) return num[k+1];
            if(k-1>=i&&num[k-1]>num[k]) return num[k];
            if(num[j]<num[k]) i = k+1;
            else if(num[i]<num[k]) j = k-1;
            else{
                num.erase(num.begin()+k);
                --j;
            }
        }
    }
};

File: java/Find_Minimum_in_Rotated_Sorted_Array_II.java
// Binary Search
// When nums[mid] = nums[low] = nums[high], we should calculate both two parts of the subarrays.
// Worst time is O(n), best time is O(logn), space is O(1).

public class Solution {
    public int findMin(int[] nums) {
        return find(nums,0,nums.length-1);
    }
    public int find(int[] nums, int low, int high) {
        while (low < high) {
            int mid = (low+high)/2;
            if (nums[mid] == nums[high]) {
                if (nums[mid] == nums[low]) {
                    return Math.min(find(nums,low,mid), find(nums,mid+1,high));
                }
                high = mid;
            }
            else if (nums[mid] < nums[high]) high = mid;
            else low = mid + 1;
        }
        return nums[low];
    }
}
